### DISCUSSION TITLE
Way to write commands

### DISCUSSION THREAD

**Discussion by jbaumgarten** [Rapptz/discord.py#8984] [Q&A]
Hi guys,

I have a code looking like the following:
```
intents = discord.Intents.default()
intents.members = True
intents.message_content = True

bot = commands.Bot(command_prefix='?', description=description, intents=intents)

@bot.event
async def on_ready():
    print(f'Logged in as {bot.user} (ID: {bot.user.id})')
    print('------')

@bot.command()
async def joined(ctx, member: discord.Member):
    """Says when a member joined."""
    await ctx.send(f'{member.name} joined {discord.utils.format_dt(member.joined_at)}')
```

I want to write the code in a class like
```
class MyClient(discord.Client):
     ...

intents = discord.Intents.default()
intents.message_content = True

client = MyClient(intents=intents)
```

I have found that @bot.event can be directly added in the class like
```
    async def on_ready(self):
        print(f'Logged in as {self.user} (ID: {self.user.id})')
        print('------')
```

But how should I proceed to add a command?
Thx :)

**Comment by CaedenPH** [comment]
@jbaumgarten 
This is a difficult question because `on_ready` is a method of `discord.Client` which you override. However, there is no such thing for commands.

I personally would recommend the use of cogs, as documented here
https://discordpy.readthedocs.io/en/stable/ext/commands/cogs.html

This way you can load cogs and put organized groups of commands within a cog

**Comment by increasingardor** [comment]
Subclass `commands.Bot` instead of `discord.Client` and you can use the command functionality built in to `Bot` already. 

If you really want to stick with `Client`, you would override the `on_message` function and manually parse messages for commands.

**Reply by CaedenPH** [reply]
> Subclass `commands.Bot` instead of `discord.Client` and you can use the command functionality built in to `Bot` already.
> 
> If you really want to stick with `Client`, you would override the `on_message` function and manually parse messages for commands.

I believe that his problem is that he wants to be able to implement commands within the `commands.Bot` subclass

**Reply by jbaumgarten** [reply]
Yes, I want to implement commands in the class MyClient directly
I'll have a look on cogs

**Reply by increasingardor** [reply]
`commands.Bot` is a subclass of `discord.Client`; everything `Client` can do `Bot` can do, with the addition of the command structure already implemented. 

What are you trying to accomplish that `Bot` doesn’t already do?

**Reply by jbaumgarten** [reply]
I'm focusing in code organisation and I'm discovering the library :)

Do you have a small example how to use commans.Bot subclass?

**Reply by increasingardor** [reply]
One of the examples here in the library:

https://github.com/Rapptz/discord.py/blob/master/examples/advanced_startup.py

**Reply by jbaumgarten** [reply]
Yes it shows how to create the custom bot but not how to create a new command inside the CustomBot class

**Reply by increasingardor** [reply]
The same way you do in a regular bot class. Freestanding commands would use the `@bot.command()` decorator (`bot` because that’s the name of their bot instance). In cogs, which are great for code organization, you would use the `@commands.command()` decorator. Bot loads the commands when you load the cog (which there’s an example of in the custom bot code above).

You can see an example of commands in cogs in this example in the library here: https://github.com/Rapptz/discord.py/blob/master/examples/basic_voice.py

Usually cogs are kept in separate files, but it works basically the same way.

**Reply by jbaumgarten** [reply]
Thx man for all your knowledge :)

What is the difference between a class heriting from commands.Bot and one from commands.Cog?

**Reply by increasingardor** [reply]
A Cog is basically a collection of Commands and Listeners. By itself it can do nothing, but a Bot can load them and add them to its list of commands and event listeners. They are mainly a method of code organization. 

For example, I have a Cog called Gifs. It contains a bunch of commands related to gifs, so we can use gifs that may not be in the built in gif picker. It’s a basic CRUD setup, with sub commands to create, retrieve, update, and delete gifs to/from database. Separately I have a Cog called Tts that manages text to speech in voice channels (we have shy users that don’t want to talk). Keeping those commands separated makes managing the files and commands much easier.

**Reply by jbaumgarten** [reply]
So you're mixing a main class heriting from discord.Client and several classes heriting from commands.Cog?

**Reply by increasingardor** [reply]
My bot class subclasses `Bot`, which itself is a subclass of `Client`. `Cog`s are basically just containers for commands that `Bot` can add to its list of commands.

**Reply by jbaumgarten** [reply]
Thx man

**Reply by jbaumgarten** [reply]
@increasingardor hi man, I made some tries but I don't understand

I created a cog class
```
class MyCogClass(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command()
    @app_commands.describe(member='The member you want to get the joined date from; defaults to the user who uses the command')
    async def joined(self, interaction: discord.Interaction, member: Optional[discord.Member] = None):
        """Says when a member joined."""
        # If no member is explicitly provided then we use the command user here
        member = member or interaction.user

        # The format_dt function formats the date time into a human readable representation in the official client
        await interaction.response.send_message(f'{member} joined {discord.utils.format_dt(member.joined_at)}')
```

I also create a bot class
```
class MyClient(commands.Bot):
    def __init__(self, command_prefix, intents: discord.Intents):
        super().__init__(command_prefix=command_prefix, intents=intents, description='Test')

    async def on_ready(self):
        logging.info(f'Logged in as {self.user} (ID: {self.user.id})')

    async def setup_hook(self):
        logging.info("Setup hook...")
        guild = discord.Object(id=MY_GUILD_ID)
        # This copies the global commands over to your guild.
        self.tree.copy_global_to(guild=guild)
        await self.tree.sync(guild=guild)
```

Then I started my app like
```
async def main():
        # Define right intents
        intents = discord.Intents.default()
        intents.members = True
        intents.guilds = True
        async with MyClient(command_prefix='/', intents=intents, db=db) as bot:
            await bot.add_cog(MyCogClass(bot))
            await bot.start('MY_TOKEN')

asyncio.run(main())
```

How ever, my bot does not display any commands "joined" on my server.
What did I miss? :/
